### PR TITLE
Collective changes from my absence

### PR DESCRIPTION
**Pull request by M-R-Schaefer**
- added feature model useful for enhanced sampling
- added simple mean energy shift
- cleaned up input sections related to model ensembles
- allow for access to last layer features in ASE calculator
- fixed erroneous multiplication with 1/sqrt(nfeatures) in gradient features
- added force features to BAL

### CODE REVIEW

File: apax/bal/feature_maps.py [hunk 1]
```diff
@@ -69,16 +69,67 @@ def inner(ll_params):
...
-            g = jnp.concatenate(g_scaled)
+
+class LastLayerForceFeatures(FeatureTransformation, extra="forbid"):
+    """
+    Model transfomration which computes the gradient of the output
+    wrt. the specified layer.
+    """
+
+    name: Literal["ll_force_feat"] = "ll_force_feat"
+    layer_name: str = "dense_2"
```
Comment (PythonFZ) [inline apax/bal/feature_maps.py:89]: Does `dense_2` assume that the apax model is always two layers deep?

File: apax/cli/templates/train_config_full.yaml [hunk 2]
```diff
@@ -44,6 +43,8 @@ model:
     r_max: 6.0
     r_min: 0.5
 
+  ensemble: null
+
```
Comment (PythonFZ) [inline apax/cli/templates/train_config_full.yaml:47]: Just for my future training runs: `n_models` has been renamed `ensemble` but is still `int|None`?

File: apax/config/train_config.py [hunk 3]
```diff
@@ -217,6 +217,22 @@ class BesselBasisConfig(BaseModel, extra="forbid"):
...
+class ShallowEnsembleConfig(BaseModel, extra="forbid"):
+    kind: Literal["shallow"] = "shallow"
+    n_members: int
+    force_variance: bool = True
+
+
+EnsembleConfig = Union[FullEnsembleConfig, ShallowEnsembleConfig]
+
+# TODO update docs
+
```
Comment (PythonFZ) [inline apax/config/train_config.py:234]: TODO

File: apax/config/train_config.py [hunk 4]
```diff
@@ -259,12 +275,14 @@ class ModelConfig(BaseModel, extra="forbid"):
...
 
+    ensemble: Optional[EnsembleConfig] = None
+
     # corrections
     use_zbl: bool = False
 
     calc_stress: bool = False
 
-    n_shallow_ensemble: int = 0
+    # n_shallow_ensemble: int = 0
```
Comment (PythonFZ) [inline apax/config/train_config.py:285]: remove?

File: apax/config/train_config.py [hunk 5]
```diff
@@ -511,7 +529,7 @@ class Config(BaseModel, frozen=True, extra="forbid"):
     n_epochs: PositiveInt
     patience: Optional[PositiveInt] = None
     seed: int = 1
-    n_models: int = 1
+    # n_models: int = 1
```
Comment (PythonFZ) [inline apax/config/train_config.py:532]: remove

File: apax/layers/readout.py [hunk 6]
```diff
@@ -15,13 +15,16 @@ class AtomisticReadout(nn.Module):
...
     n_shallow_ensemble: int = 0
+    is_feature_fn: bool = False
     dtype: Any = jnp.float32
 
     def setup(self):
-        readout_unit = [1]
-        if self.n_shallow_ensemble > 0:
-            readout_unit = [self.n_shallow_ensemble]
-        units = [u for u in self.units] + readout_unit
+        units = [u for u in self.units]
```
Comment (PythonFZ) [inline apax/layers/readout.py:22]: Why not `list(self.units)` and what type is `self.units` that it has to be iterated trough like this?

Comment (PythonFZ) [reply apax/layers/readout.py:22]: Ahh, I see you make a copy

File: apax/model/builder.py [hunk 7]
```diff
@@ -155,14 +166,43 @@ def build_energy_derivative_model(
...
             model = EnergyDerivativeModel(
                 energy_model,
                 calc_stress=self.config["calc_stress"],
             )
         return model
+
+    def build_ll_feature_model(
+        self,
+        scale=1.0,
+        shift=0.0,
```
Comment (PythonFZ) [inline apax/model/builder.py:191]: These are not used, are they?

File: apax/model/gmnn.py [hunk 8]
```diff
@@ -51,9 +52,13 @@ def __call__(
 class FeatureModel(nn.Module):
     """Model wrapps some submodel (e.g. a descriptor) to supply distance computation."""
 
-    feature_model: nn.Module = GaussianMomentDescriptor()
+    # atomistic_model: nn.Module = GaussianMomentDescriptor()
```
Comment (PythonFZ) [inline apax/model/gmnn.py:55]: remove

File: apax/model/gmnn.py [hunk 9]
```diff
@@ -75,7 +80,14 @@ def __call__(
             perturbation,
         )
 
-        features = self.feature_model(dr_vec, Z, idx)
+        # features = self.atomistic_model(dr_vec, Z, idx)
```
Comment (PythonFZ) [inline apax/model/gmnn.py:83]: remove

**Review by PythonFZ — COMMENTED** [review]
I just briefly read over it and left some comments. Looking forward to test the shallow ensembles. I think the `yaml` example files should be a bit clearer with respect to the ensemble / the ensemble type should be documented somewhere.

File: apax/bal/feature_maps.py (hunk 1 above)
Comment (M-R-Schaefer) [reply apax/bal/feature_maps.py:89]: that's just the default, it's user supplied. It has to be one of the linear layers though the way it is currently implemented. But it's already possible (also in IPSuite) to choose an arbitrary linear layer.
Intermediate layers don't really make sense to use. I could code it in a way that always selects the last linear layer automatically, but that's pretty low on my priority list atm.

File: apax/model/builder.py (hunk 7 above)
Comment (M-R-Schaefer) [reply apax/model/builder.py:191]: no. This was to keep the signature in line with the other model building methods. But I guess it's not needed.

File: apax/bal/feature_maps.py (hunk 1 above)
Comment (PythonFZ) [reply apax/bal/feature_maps.py:89]: Okay, then this is perfectly fine.

File: apax/model/builder.py (hunk 7 above)
Comment (PythonFZ) [reply apax/model/builder.py:191]: maybe add a comment and set them to None?

File: apax/config/train_config.py [hunk 10]
```diff
@@ -217,6 +217,22 @@ class BesselBasisConfig(BaseModel, extra="forbid"):
...
+class FullEnsembleConfig(BaseModel, extra="forbid"):
+    kind: Literal["full"] = "full"
+    n_members: int
+
+
+class ShallowEnsembleConfig(BaseModel, extra="forbid"):
+    kind: Literal["shallow"] = "shallow"
+    n_members: int
+    force_variance: bool = True
+
```
Comment (Tetracarbonylnickel) [inline apax/config/train_config.py:229]: Docs are missing, otherwise config is not explained.

**Review by Tetracarbonylnickel — CHANGES_REQUESTED** [review]

File: apax/cli/templates/train_config_full.yaml (hunk 2 above)
Comment (M-R-Schaefer) [reply apax/cli/templates/train_config_full.yaml:47]: no, ensemble lets' you choose the kind of ensemble, hoe many members and in the case of shallow ensembles, whether or not to compute force variance.

**Review by Tetracarbonylnickel — COMMENTED** [review]
are the examples still working? I see no changes there and a lot of changes in the configs. Can imagine that this causes problems.

**Comment by M-R-Schaefer** [comment]
> are the examples still working? I see no changes there and a lot of changes in the configs. Can imagine that this causes problems.

the only part of the config which has changed is the ensemble section and we don't use ensembles in the examples.
I ran them all again and they work without modification.

**Review by Tetracarbonylnickel — APPROVED** [review]
looks good for me.
